src/history/hs_cursor.c

```diff
@@ -409,24 +409,24 @@ __wt_hs_find_upd(WT_SESSION_IMPL *session, WT_ITEM *key, const char *value_forma
 
     __wt_free_update_list(session, &mod_upd);
     while (modifies.size > 0) {
-        __wt_modify_vector_pop(&modifies, &upd);
-        __wt_free_update_list(session, &upd);
+        __wt_modify_vector_pop(&modifies, &mod_upd);
+        __wt_free_update_list(session, &mod_upd);
```

Comment:
Can't we fix the WT_TRET while we're in the area? It could cause the same problem if it fails.

Reply:
I moved TRET to the end of the function. Does that fix the issue?

Reply:
Also I added a change to mark the buffer as invalid if we encounter any error or the value is not found.

Reply:
I think marking the buffer as invalid seems like the right solution.

Why does the caller ignore WT_NOTFOUND, anyway? Is there a path where the current choice is good, even if history-store finds nothing?

Reply:
I don't think there is such a path. It is just how it is implemented currently. The caller depends on the invalid type not the return values to decide whether something is found. I think there is any underlying assumption here in the code. If we return WT_NOTFOUND, then the buffer type must be invalid. I feel this invalid state is redundant if we can honour the WT_NOTFOUND return type. I'll try to make a change to get rid of it.

Reply:
mhh, I give up on that as I found the caller code in the search logic in a lot of places assumes WT_NOTFOUND is not an issue and should return 0. Such change is too risky.

Reply:
After reading the code, I think not return WT_NOTFOUND from `__wt_hs_find_upd ` is the appropriate decision. WT_NOTFOUND is the return of cursor API that indicates the user finds nothing. However, here we are at btree level, which is inside a cursor API call. The underlying btree search logic has placed us on a btree record and we are here to decide whether it is visible or not. If we ought to return an error to tell the caller nothing is visible, we should create a new error code, i.e., WT_NOTVISIBLE instead of using WT_NOTFOUND. As in the upper level cursor logic, it may continue to walk to the next btree record if the current one is not visible and it is very easy to cause confusion.